### PR TITLE
[WIP] v1.0 refactor 

### LINKED ISSUE #99: Simplify configurations override
Use config.js in src directory that will merge default configs with particular config file

### LINKED ISSUE #758: Only one foil card displayed at a time
Follow up to #729 (was addressed in #736)

When foil cards are picked from a pack, only the last picked foil gets the foil effect applied. All former ones loose the effect and are displayed as normal cards.
It doesn't matter if column view is enabled or disabled.

### LINKED ISSUE #811: Add unit test for supported json structure
With #776 XML upload got added and there is a test case for it.

There should be a test case for json as well.

### LINKED ISSUE #828: Coldsnap lands
The Coldsnap packs don't include the snow basics from the set, making many cards unplayable. 
Is there a good fix to this?

### LINKED ISSUE #833: "Game over" error aborting draft
Firstly, thank you for an amazing website. 

Yesterday my friends and I were cube drafting. We were 4 people in total, drafting 5 packs, with the "leisurely" time setting. Somewhere in the middle of the fifth and last pack all of us were instantly booted from the draft with a red "game over" error text in red. I'll attach a screenshot. 

Obviously we have no idea what caused this, but speculate the draft might have some hidden time limit or similar? I'd say we were pretty much exactly an hour in when the error hit us. 

Would this be possible to fix? Is there any way we can avoid this in the future other than simply drafting more quickly? Thanks again! 

IMG_20200118_121903

### PR DESCRIPTION
**Pull request by HerveH44**
This is my attempt to refactor mostly backend and also frontend. 

The main idea behind this refactor is to avoid specific work on mtgjson file parsing and booster generation. Actually I'm using a set of rules written by @taw that explain how to create a booster (checking for specific rules for each set, balance colors and include foil if possible). 

The second idea is to apply "functional programming" style especially to ease tests. I'm working on including tests as much as possible when it's needed (spec files) so we can know when we break something important (at least on the backend).
 
Todo List:
- [x] Improve README
    - [x] Explain how the game works (EngineIO, game options..., booster generation)
    - [x] Add development section 
    - [x] Add test section
- [ ] Add as much documentation as possible
    - [x] Add API documentation
    - [ ] Add backend and frontend documentation (through JSDocs maybe or just simple markdowns)
- [x] Add more tests on import the cards and generate boosters
    - [x] Use specific cards / sets (double face cards, devoid, old sets, ...)
- [x] Rename main folders (frontend & backend seems better than public & src. @ZeldaZach do you agree?)
- [x] Rename folder "src/make" to "src/import"
- [x] Organize scripts folder and document it on README
    - [x] Move scripts like "update_database" to scripts folder 
- [x] Merge master branch
- [x] Merge config branch
- [x] Use only const and let everywhere instead of var
- [x] Replace internal librairies with Lazy.js and Chance.js
- [x] Delete score -> bots will always randomly pick

closes #99, closes #758, closes #811 , closes #833, closes #828

### CODE REVIEW

**Comment by HerveH44** [comment]
@tooomm and/or anyone could you try to test this PR? I'd like to know if you notice anything strange. Most of the things should work normally. 

I notices that it's a bit slower (because the booster generation is based on probability which implies more computation but I still have work to do on this part) but everything (cube, normal and chaos) should work as expected. 

Thanks in advance,

**Comment by ZeldaZach** [comment]
I’ll check this out soon. Great job!

**Comment by taw** [comment]
Hi guys, I setup automated export to https://github.com/taw/magic-sealed-data
I hope it helps.

**Comment by tooomm** [comment]
Are you sure foil cards are still correctly inserted @HerveH44?
I tested a few drafts with bots, but did not saw a single foil card.
Or did the probability drastically change? On dr4ft you do 1-2 drafts with bots and you see several foils normally.

Where did you took @taw's rules from? Is there a source?
What is the different logic/improvement to our former method? Any new features or changed probabilities?

**Comment by taw** [comment]
@tooomm JSON is generated by code in https://github.com/taw/magic-search-engine

To get started [this](https://github.com/taw/magic-search-engine/blob/master/search-engine/lib/card_sheet_factory.rb) and [this](https://github.com/taw/magic-search-engine/blob/master/search-engine/lib/pack_factory.rb) files.

Research to get those rules was long and tedious. I'm sure it's not completely accurate, but it's probably a lot more accurate than any other rules, and issues/PRs welcome, it's all Open Source.

**Comment by tooomm** [comment]
I found some foil cards now, they work! :)

However, I had two foils of the same card in one pack at one point:
double foil

Maybe all cards of the same name are made foil instead only one?
Can it happen, that there is 2x the same card in one pack if one is a foil? I don't recall if that's possible.

@taw Do you have the numbers and some logic regarding foil cards as well? How do you do it for your generator? I found some numbers in a quick search, and there seem to be some special cases as well in older sets regarding appearance and what they replace:
https://mtg.gamepedia.com/Foil_card#Rarity
https://www.mtgsalvation.com/forums/magic-fundamentals/magic-general/325533-how-do-booster-pack-foils-work

**Comment by HerveH44** [comment]
@tooomm I think that's a bug from frontend. I'm pushing something that should fix it, not quite sure as it's very hard to reproduce.

**Comment by taw** [comment]
> @taw Do you have the numbers and some logic regarding foil cards as well? How do you do it for your generator? I found some numbers in a quick search, and there seem to be some special cases as well in older sets regarding appearance and what they replace

Unfortunately no official information.  Some comments with links in the code:

>Based on https://www.reddit.com/r/magicTCG/comments/snzvt/simple_avr_sealed_simulator_i_just_made/c4fk0sr/
> Details probably vary by set and I don't have too much trust in any of these numbers anyway

And:

> We don't have anywhere near reliable information
> Masterpieces supposedly are in 1/144 booster (then 1/129 for Amonkhet), and they're presumably equally likely
> 
> These numbers could be totally wrong. I base them on a million guesses by various internet commenters.
> 
> Maro says basic foils and common foils are equally likely [https://twitter.com/maro254/status/938830320094216192]

We can pretty safely assume that each specific common = each specific basic, and each specific rare = 2x each specific mythic. That masterpiece rate and total foil rate come from some official sources.

That leaves common : uncommon : rare ratio as a guess, but various internet sources give similar numbers.

It's of course also totally possible that I made some mistakes. If you see something wrong with my code, let me know.

**Comment by tooomm** [comment]
@HerveH44 Currently, the site is only blank after loading up.

**Comment by HerveH44** [comment]
@tooomm should it good for tests. I'm not going to do any more refactor, just adding some tests. 

If you find any problem tell me!

**Comment by ZeldaZach** [comment]
This is bugging out. I just did a draft and all of my cards I picked turned into other cards in the sideboard. It says I draft 10 hushbringers and 30 Vantress Paladins...

**Comment by ZeldaZach** [comment]
I _really_ don't like the random bot picking. I drafted 20 of 24 rares in a test. We need to discuss bot draft practices after the rewrite is done before we deploy as 1.0

**Comment by tooomm** [comment]
@zeldazach It used to always be like that as the bot data was only for old sets.
Maybe a bit less stupid bots are an option. With general rules instead particular card/set knowledge.
Like stick to colors mostly at one point. 

But how to get started? Pick randomly in first 3 rounds with an emphasis on rarer cards? And once a Color appears again try to stick to that with like 90% chance. Also, count how many cards of a Color you are seeing in packs = likelihood that they wheel for example?
Consider rarity, especially mostly don’t pass on-color rares for example.
Or even take creature count into account as well as.
When no on-color pick available. Randomly pick again...
There is a lot to think of! :/

**Comment by HerveH44** [comment]
I applied couple fixes. That should stop bugging out.

File: backend/import/toBoosterCard.spec.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  describe("can parse special cards", () => {
+    it("parse a devoid card as color is colorless", () => {});
+    it("parse a DFC card as doubleFaced and with a flippedCardURL", () => {});
+    it("parse a transform card as doubleFaced and with a flippedCardURL", () => {});
+    it("parse a split|aftermath|adventure card with double name", () => {});
+    it("parse a split card with correct mana cost", () => {});
+    it("parse a multicolor card as multicolor", () => {});
+    it("parse a no color card as colorless", () => {});
+  });
```
[the remark this answers is not in the document]
Comment (HerveH44) [reply backend/import/toBoosterCard.spec.js]: Great! Could you tell me all the corner cases that come to your mind so I can put them as tests? If you can bring card names for each situation that would be easier to test.

Thanks

**Comment by HerveH44** [comment]
@tooomm my answers to your points.

    #811 not addressed yet as far as I can see.
==> Not yet no. I don't know how to check for errors as I'm expecting a MTGJson.

    As you incorporated #99 here, I assume your separate config branch is no longer needed in the repo? Just want to double check before deleting it.
==> wait until the PR is merged before doing it. I merged from it so I don't want to make any problem because of it.
    The sorting options do only get applied to the next draft pack that is passed to you, the current one is no longer affected
==> fixed the pb
    Download option should only be available at the end of the draft (grey button out until all picks are made?). Name and file type should be changable before? Or just make it all available at the end? It could also be connected to the deck size option and only offers a download if the amount of cards in the deck match those number to avoid problems. Maybe that's something for it's own ticket and some more discussion.
==> that's the behavior of dr4ft right now. I change is needed ( and I don't think it does ) that should be discussed in a separate issue.
    Same for copy deck to clipboard button
    Downloading a deck should creates a hash, if you make changes and download again, the hash should update again. That works in sealed, but not in draft.
==> I don't reproduce your problem.

    Cube Sealed only allows up to 30 cards per player anymore
==> fixed the pb
    ESLint check errors at travis
==> Don't see any errors.

    Selected Autopick doesn't work for last card in pack in the last pack
==> fixed the pb

**Comment by tooomm** [comment]
@HerveH44 

> Download option should only be available at the end of the draft (grey button out until all picks are made?). Name and file type should be changable before? Or just make it all available at the end? It could also be connected to the deck size option and only offers a download if the amount of cards in the deck match those number to avoid problems. Maybe that's something for it's own ticket and some more discussion.
> ==> that's the behavior of dr4ft right now. I change is needed ( and I don't think it does ) that should be discussed in a separate issue.
>
> Same for copy deck to clipboard button

Will open a new issue 👍 
<br>

> Downloading a deck should creates a hash, if you make changes and download again, the hash should update again. That works in sealed, but not in draft.
> ==> I don't reproduce your problem.

During the draft it keeps generating new hashes for me too when I download or copy the deck, but at the end of the draft when I build a final deck and add (suggested) lands, it doesn't work anymore.

<br>

> ESLint check errors at travis
> ==> Don't see any errors.

See https://travis-ci.com/dr4fters/dr4ft/builds/144810003, mostly warnings though.

File: backend/import/toBoosterCard.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tooomm) [reply backend/import/toBoosterCard.spec.js]: Some examples at least:

fuse: [Down / Dirty](https://scryfall.com/card/dgm/126/down-dirty)
meld: [Gisela, the Broken Blade](https://scryfall.com/card/emn/28a/gisela-the-broken-blade?utm_source=mw_MTGWiki)
flip: [Akki Lavarunner](https://scryfall.com/card/chk/153/akki-lavarunner-tok-tok-volcano-born?utm_source=mw_MTGWiki)
split with more faces: [Who/What/When/Where/Why](https://scryfall.com/card/unh/120/who-what-when-where-why)

**Comment by HerveH44** [comment]
> During the draft it keeps generating new hashes for me too when I download or copy the deck, but at the end of the draft when I build a final deck and add (suggested) lands, it doesn't work anymore.

@tooomm I'm sorry I don't reproduce that. Tried at the end of the draft too and it still produces the different hashses

**Comment by tooomm** [comment]
> > During the draft it keeps generating new hashes for me too when I download or copy the deck, but at the end of the draft when I build a final deck and add (suggested) lands, it doesn't work anymore.
> 
> @tooomm I'm sorry I don't reproduce that. Tried at the end of the draft too and it still produces the different hashses

For me, downloading a deck after a draft finishes creates no hash at all. Even adding lands or doing other changes doesn't work. I can still download it.

<br>

>>Selected Autopick doesn't work for last card in pack in the last pack
>
>==> fixed the pb

It might pick the card now and add it to your pool, but the picking area "Pack x - Pick y/z" still stays and shows that last card marked as "Autopick" even after the time run off. It freezes. You still have to manually click the card for the second time (aka manually pick) to make that section and the marked card disappear.

**Comment by tooomm** [comment]
Status: https://github.com/dr4fters/dr4ft/pull/798/#pullrequestreview-344101924

**Comment by ZeldaZach** [comment]
Thinking of ideas on how we can handle drafts better. https://aetherhub.com/Apps/LimitedRatings goes back quite a while, so we could use their draft stats number to make better informed choices for the bots.

Rudimentary idea I'm having on how to make a better draft bot, since we don't have a good starting stat base. Would love to try and partner with someone to get more data.

if set exists on AetherHub:
- First three picks of draft, take highest ranking card rank
- After those picks, select highest ranking card that falls within the same colors as the first three

if set does not exist on AetherHub:
- First pick the rare
- Pick cards that fall in the same color ranges as other cards drafted

if cube draft:
- Pick randomly at all times

**Comment by HerveH44** [comment]
@ZeldaZach I think the bots picking strategy should be another issue. This PR is just about refactoring the backend, not implementing anything new.

**Comment by HerveH44** [comment]
@tooomm you can test now, I solved the 2 problems about endgame (autopick and hash)

**Comment by tooomm** [comment]
@HerveH44 Sorry for the delay - it looks good! :)
Only #811 is left, or remove the "closes" reference from the starting post maybe?

**Comment by ZeldaZach** [comment]
For some reason I keep getting beeps even when I tell it I want desktop notifications instead. That and the robot "AI" needing work, are my only complaints

**Comment by tooomm** [comment]
Bots are their own issue now as that should not be part of this PR @ZeldaZach 

Beeping I will check again

**Comment by ZeldaZach** [comment]
I'm happy with how this turned out. It's a good start into the V1 future, and gives a base for future PRs to be on. Will merge and deploy.
